EXPORTING_AND_IMPORTING_DATABASE.md

```diff
@@ -0,0 +1,131 @@
+   > ckan_datastore-`date -I`.sql
+```
+
+Log out of the remote server.
+
+Copy the files to your local machine:
+```
+scp ubuntu@opendataphilly.org:database_dumps/* .
+```
+
+Delete the export files from the server:
+```
+ssh ubuntu@opendataphilly.org "rm -r database_dumps"
+```
+
+Start a Postgres 9.4 docker container:
+```
+docker run --name odp_export -e POSTGRES_PASSWORD=odp -d postgres:9.4
+```
+
+Create and import the main database:
+```
+docker run -i --rm -e PGPASSWORD=odp --link odp_export:postgres postgres:9.4 \
+   psql -h postgres -U postgres postgres -c "CREATE DATABASE ckan_default;"
+docker run -i --rm -e PGPASSWORD=odp --link odp_export:postgres postgres:9.4 \
+   psql -h postgres -U postgres postgres -c "CREATE USER ckan_default;"
+docker run -i --rm -e PGPASSWORD=odp --link odp_export:postgres postgres:9.4 \
+   psql -h postgres -U postgres ckan_default < ckan_default-`date -I`.sql
+```
+
+Make anonymizing queries.
+You'll need the `faker` python package, so either make a virtualenv, activate it, and run `pip install faker` or just install it globally.
+Run the command and write the resulting queries to a file:
+```
+./generate_anon_queries.sh > anon_queries.sql
+```
+
+Run the queries against the main database:
+```
+docker run -i --rm -e PGPASSWORD=odp --link odp_export:postgres postgres:9.4 \
+   psql -h postgres -U postgres ckan_default < anon_queries.sql
+```
+
+Export the sanitized database
+```
+docker run -i --rm -e PGPASSWORD=odp --link odp_export:postgres postgres:9.4 \
+   pg_dump -h postgres -U postgres -d ckan_default > ckan_default_sanitized-`date -I`.sql
+```
+
+Now that you have the sanitized version, you can delete the non-sanitized `ckan_default` file and stop and remove the database container:
+```
+rm ckan_default-`date -I`.sql
+docker stop odp_export && docker rm -v odp_export
+```
+
+Since SQL files compress quite a bit, put the two files into a tarball:
+```
+tar czvf ODP_db_dumps-`date -I`.tar.gz \
+    ckan_default_sanitized-`date -I`.sql ckan_datastore-`date -I`.sql
+```
+
+Now copy the `.tar.gz` file to a suitable storage location and, if you're done with them, delete the separate files.
+
+
+### Importing into development
+
+Extract the export files into your project root (or somewhere else that's within a directory that's mapped into the VM).
+
+Import them (fill in the filenames and, if you put them in a different directory, adjust the paths):
+```
+vagrant ssh database<<EOF
+export PGPASSWORD=ckan_default
+psql -U ckan_default -h 127.0.0.1 -d ckan_default -f /vagrant/ckan_default_sanitized-YYYY-MM-DD.sql
+psql -U ckan_default -h 127.0.0.1 -d datastore_default -f /vagrant/ckan_datastore-YYYY-MM-DD.sql
+EOF
+```
+
+That should print a bunch of SQL messages.
+
+Restart the services:
+```
+vagrant ssh app <<EOF
+sudo service apache2 restart
+sudo service jetty restart
+sudo ckan search-index rebuild -r
+sudo ckan views create -y
```

Comment:
I followed the testing instructions, but when importing the latest DB dumps, this command returned the following error:
```
2018-11-20 19:46:06,197 INFO  [ckan.lib.cli] No view types provided, using default types
Traceback (most recent call last):
  File "/usr/bin/ckan", line 45, in <module>
    load_entry_point('PasteScript', 'console_scripts', 'paster')()
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/paste/script/command.py", line 102, in run
    invoke(command, command_name, options, args[1:])
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/paste/script/command.py", line 141, in invoke
    exit_code = runner.run(args)
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/paste/script/command.py", line 236, in run
    result = self.command()
  File "/usr/lib/ckan/default/src/ckan/ckan/lib/cli.py", line 2146, in command
    self.create_views(view_plugin_types)
  File "/usr/lib/ckan/default/src/ckan/ckan/lib/cli.py", line 2380, in create_views
    view_types=loaded_view_plugins)
  File "/usr/lib/ckan/default/src/ckan/ckan/lib/datapreview.py", line 303, in add_views_to_dataset_resources
    create_datastore_views)
  File "/usr/lib/ckan/default/src/ckan/ckan/lib/datapreview.py", line 243, in add_views_to_resource
    context, {'id': resource_dict['id']})
  File "/usr/lib/ckan/default/src/ckan/ckan/logic/__init__.py", line 464, in wrapped
    result = _action(context, data_dict, **kw)
  File "/usr/lib/ckan/default/src/ckan/ckan/logic/action/get.py", line 1137, in resource_view_list
    raise NotFound
ckan.logic.NotFound
```

Reply:
I just realized the testing instructions above were kind of garbled, with step 2 incomplete and step 3 redundantly saying to provision `app`.  I edited them.  Were you running those service restart  commands with a fully provisioned `app` instance?  The error message makes me wonder if there's a plugin or something that's activated but not installed.

Reply:
I _did_ run `vagrant provision app` before hand, and I tried that again and got the same error.

I then deleted my `app` VM and recreated it, and I was able to make it through the import instructions when I tried again. So.... :man_shrugging: